### DISCUSSION TITLE
Is it possible to author a driver for a database that does not support auto-incrementing column types?

### DISCUSSION THREAD

**Discussion by tmc** [ent/ent#3166] [General]
Some databases do not support auto-incrementing column types, Cloud Spanner comes to mind.

Could you provide some direction on the right way to write/support this type of database driver? I presume some extension of `sqlgraph` and an optionally implemented interface (or two) would be involved.

Please advise.

**Comment by crossworth** [comment]
Hello @tmc, I'm not sure if I understood your question, Ent has a lot of hooking points, the [Driver](https://github.com/ent/ent/blob/dcc16c5763eb58af917b1c780f196597c86beba8/dialect/sql/driver.go) is one of them, but is relative simple and most of the time only delegates to the `database/sql` driver.

What you probably want is to add another [dialect](https://github.com/ent/ent/blob/dcc16c5763eb58af917b1c780f196597c86beba8/dialect/dialect.go#L19-L22) if the database is not compatible with the wire protocol  of the current supported databases. You may have to add a few checks to the [Builder](https://github.com/ent/ent/blob/dcc16c5763eb58af917b1c780f196597c86beba8/dialect/sql/builder.go#L592-L606) and last you will have to add support for the [schema](https://github.com/ent/ent/tree/dcc16c5763eb58af917b1c780f196597c86beba8/dialect/sql/schema), I know that you are adding support for [Spanner on Atlas](https://github.com/ariga/atlas/tree/spanner), so this should be quite easy for you.

One way to visualize the operations (like Query/Insert/Update/Delete) is like this (not quite right, there are a few hooks here and there and few parts were omitted):
![image](https://user-images.githubusercontent.com/1251151/206866394-f44652d0-ec30-41ef-a9da-d4a8227c23b7.png)
_image generated using https://github.com/terrastruct/d2_

Now, about the question related to auto-incrementing, Google has an [article talking about it](https://cloud.google.com/solutions/sequence-generation-in-cloud-spanner), and [another article talking](https://cloud.google.com/spanner/docs/migrating-mysql-to-spanner#sequences) about how migrate from MySQL.

I would like to quote a few parts from the articles:
> In fact, using monotonically increasing (or decreasing) values as row keys does not follow [best practices](https://cloud.google.com/spanner/docs/schema-design#choose_a_primary_key_to_prevent_hotspots) in Spanner because it creates hotspots in the database, leading to a reduction in performance.

> using monotonically increasing numbers as primary keys is an anti-pattern in Cloud Spanner. An alternative mechanism of generating a unique primary key is to use a [random UUID](https://wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)).

With that in mind, I guess one of the solutions would be to disallow auto-increment and only support custom IDs (like UUID, XID, ULID).

Since Cloud Spanner itself don't impose an ID type, it's hard to say what the best option would be.

Also worth reading: [Schema design best practices](https://cloud.google.com/spanner/docs/schema-design#primary-key-prevent-hotspots).

With all that said, if you really want to implement auto increment, one solution would to have a table to control and generate the ids, Cloud Spanner don't seem to support stored procedures, so you will have to do this on the application side, maybe you can use [Ent hooks](https://entgo.io/docs/hooks/) or implement as custom ID type.

**Reply by JoshGlazebrook** [reply]
@tmc Are you still working on this?

**Reply by tmc** [reply]
@JoshGlazebrook no not actively.
